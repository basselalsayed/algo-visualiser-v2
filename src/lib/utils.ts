import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type T0 =
  | 'string'
  | 'number'
  | 'bigint'
  | 'boolean'
  | 'symbol'
  | 'undefined'
  | 'object'
  | 'function';

export function assert(value: unknown, type: T0): asserts value is T0 {
  const actualType = typeof value;

  if (type === 'object') {
    if (value === null) {
      throw new TypeError(`Expected type '${type}', but received null`);
    }
  }

  if (actualType !== type) {
    throw new TypeError(
      `Expected type '${type}', but received '${actualType}'`
    );
  }
}

export function typeIs(value: unknown, type: T0): value is T0 {
  try {
    assert(value, type);
    return true;
  } catch {
    return false;
  }
}

export function convertToMilliseconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const milliseconds = end - start;
  return Math.floor(milliseconds * 1000) / 1000; // 3 decimal places
}

export function convertToSeconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const seconds = (end - start) / 1000;
  return Math.floor(seconds * 1000) / 1000;
}

export function elementIsFullyScrolled(element: HTMLElement): boolean {
  return (
    Math.floor(element.scrollHeight - element.scrollTop) ===
    Math.floor(element.clientHeight)
  );
}
