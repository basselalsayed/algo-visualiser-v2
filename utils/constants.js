import path from 'node:path';

export const DIRECTORIES = [
  'algorithms',
  'components',
  'contexts',
  'data',
  'hooks',
  'lib',
  'store',
];

export const ROOT_PATH = path.resolve(import.meta.dirname, '..');
