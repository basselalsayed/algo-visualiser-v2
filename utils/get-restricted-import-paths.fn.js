import fs from 'node:fs';
import path from 'node:path';

const __dirname = new URL('.', import.meta.url).pathname;

/**
 * Recursively walks through a directory and collects relative paths to all index.ts or index.tsx files.
 *
 * @param {string} dir - Absolute path to start walking from.
 * @param {string} baseRelativePath - Relative path from the module root (used for nested structure).
 * @returns {string[]} - List of relative paths to directories containing index files.
 */
function findIndexFileDirs(dir, baseRelativePath = '') {
  const indexDirs = [];
  const entries = fs.readdirSync(dir);

  for (const entry of entries) {
    const fullPath = path.join(dir, entry);
    const relPath = path.join(baseRelativePath, entry);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      indexDirs.push(...findIndexFileDirs(fullPath, relPath));
    } else if (
      (entry === 'index.ts' || entry === 'index.tsx') &&
      baseRelativePath !== ''
    ) {
      indexDirs.push(baseRelativePath);
    }
  }

  return indexDirs;
}

/**
 * Generates a list of restricted import paths for all index.ts(x) files in a module.
 *
 * @param {string} dirAbsolutePath - Absolute path to the module directory (e.g. src/components)
 * @param {string} alias - Import alias for the module (e.g. @/components)
 * @returns {Array<{ name: string, message: string }>}
 */
function getRestrictedIndexImports(dirAbsolutePath, alias) {
  const indexDirs = findIndexFileDirs(dirAbsolutePath);

  const staticUpwardPaths = [
    '../../../index',
    '../../..',
    '../../index',
    '../..',
    '../index',
    '..',
  ].map((name) => ({
    message: `Avoid importing '${name}'. Use aliased or direct file paths.`,
    name,
  }));

  const indexPathViolations = indexDirs.flatMap((relPath) => [
    {
      message: `Avoid importing index file: ${alias}/${relPath}/index.ts. Import specific files instead.`,
      name: `${alias}/${relPath}`,
    },
    {
      message: `Avoid importing index file: ../${relPath}/index.ts. Import specific files instead.`,
      name: `../${relPath}`,
    },
  ]);

  return [...staticUpwardPaths, ...indexPathViolations];
}

export { __dirname, getRestrictedIndexImports };
