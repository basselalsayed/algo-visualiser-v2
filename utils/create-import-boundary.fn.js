import path from 'node:path';

import { DIRECTORIES, ROOT_PATH } from './constants.js';
import { getRestrictedIndexImports } from './get-restricted-import-paths.fn.js';

/**
 * @param {string} dir
 */
const getAlias = (dir) => `@/${dir}`;

/**
 * @param {string} dir
 */
export function createImportBoundary(dir) {
  const alias = getAlias(dir);
  const glob = `src/${dir}/**/*.{ts,tsx}`;

  const forbid = DIRECTORIES.flatMap((_dir) =>
    _dir === dir ? undefined : [`${getAlias(_dir)}/**/*`, `${getAlias(_dir)}/*`]
  ).filter(Boolean);

  const absoluteDir = path.resolve(ROOT_PATH, 'src', dir);

  const restrictedIndexPaths = getRestrictedIndexImports(absoluteDir, alias);

  return {
    files: [glob],
    rules: {
      'import/no-internal-modules': [
        'error',
        {
          forbid,
        },
      ],
      'no-restricted-imports': [
        'error',
        {
          paths: [
            ...restrictedIndexPaths,
            {
              message: `Avoid importing '${alias}' from within ${dir}/. Use direct file paths.`,
              name: alias,
            },
          ],
          patterns: [
            {
              group: ['**/index', '**/index.ts', '**/index.tsx'],
              message: `Avoid importing index files directly. Import specific files instead.`,
            },
            {
              group: ['../../**/*'],
              message: `Avoid relative imports outside of this module. Use alias paths instead.`,
            },
          ],
        },
      ],
    },
  };
}
