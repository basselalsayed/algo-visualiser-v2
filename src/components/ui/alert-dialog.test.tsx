import { renderDialogSnapshot } from '@/test/utils';

import {
  AlertDialog,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
} from './alert-dialog';

describe('AlertDialog.component', () => {
  it('looks the same', () => {
    expect(
      renderDialogSnapshot(
        <AlertDialog open>
          <AlertDialogContent>
            <AlertDialogHeader>
              <AlertDialogTitle>Title</AlertDialogTitle>
              <AlertDialogDescription>
                DesAlertDialogDescription
              </AlertDialogDescription>
            </AlertDialogHeader>

            <AlertDialogFooter>
              <AlertDialogCancel />
              <AlertDialogAction />
            </AlertDialogFooter>
          </AlertDialogContent>
        </AlertDialog>
      )
    ).toMatchSnapshot();
  });
});
