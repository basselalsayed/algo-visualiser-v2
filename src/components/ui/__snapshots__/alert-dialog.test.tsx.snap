// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlertDialog.component > looks the same 1`] = `
<body
  data-scroll-locked="1"
  style="pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />
  <div
    aria-hidden="true"
    class="fixed inset-0 z-50 bg-black/50 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:animate-in data-[state=open]:fade-in-0"
    data-aria-hidden="true"
    data-slot="alert-dialog-overlay"
    data-state="open"
    style="pointer-events: auto;"
  />
  <div
    aria-describedby="radix-«r2»"
    aria-labelledby="radix-«r1»"
    class="fixed top-[50%] left-[50%] z-50 grid w-full max-w-[calc(100%-2rem)] translate-x-[-50%] translate-y-[-50%] gap-4 rounded-lg border bg-background p-6 shadow-lg duration-200 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95 data-[state=open]:animate-in data-[state=open]:fade-in-0 data-[state=open]:zoom-in-95 sm:max-w-lg"
    data-slot="alert-dialog-content"
    data-state="open"
    id="radix-«r0»"
    role="alertdialog"
    style="pointer-events: auto;"
    tabindex="-1"
  >
    <div
      class="flex flex-col gap-2 text-center sm:text-left"
      data-slot="alert-dialog-header"
    >
      <h2
        class="font-semibold"
        data-slot="alert-dialog-title"
        id="radix-«r1»"
      >
        Title
      </h2>
      <p
        class="text-sm text-muted-foreground"
        data-slot="alert-dialog-description"
        id="radix-«r2»"
      >
        DesAlertDialogDescription
      </p>
    </div>
    <div
      class="flex flex-col-reverse gap-2 sm:flex-row sm:justify-end"
      data-slot="alert-dialog-footer"
    >
      <button
        class="inline-flex items-center justify-center gap-2 rounded-md text-base font-medium whitespace-nowrap transition-colors focus-visible:ring-1 focus-visible:ring-ring focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 h-9 px-4 py-2 animate_grad bg_grad_accent--outline--text shadow-xs"
        type="button"
      />
      <button
        class="inline-flex items-center justify-center gap-2 rounded-md text-base font-medium whitespace-nowrap transition-colors focus-visible:ring-1 focus-visible:ring-ring focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 h-9 px-4 py-2 animate_grad bg_grad_accent border-0 shadow-xs"
        type="button"
      />
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
